docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+# TiDB Design Documents
+
+- Author(s): [morgo](http://github.com/morgo)
+- Discussion PR: https://github.com/pingcap/tidb/pull/30558
+- Tracking Issue: https://github.com/pingcap/tidb/issues/30366
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Test Design](#test-design)
+* [Impacts & Risks](#impacts--risks)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+Currently, TiDB has two primary methods of configuration:
+
+- The configuration file (in `.toml` format) ([Docs](https://docs.pingcap.com/tidb/stable/tidb-configuration-file))
+- Using MySQL compatible system variables (`SET GLOBAL sysvar=x`, `SET SESSION sysvar=x`) ([Docs](https://docs.pingcap.com/tidb/stable/system-variables))
+
+Both the **semantics** and **naming conventions** differ between these two methods:
```

Comment:
Could you supply a real user scenario that would require this *non-persisted* instance scoped variables? From my perspective, setting a variable in instance level but not persisting it will lead to confusion in many cases. For example, if I really would like to enable a feature for only one instance, I would be very surprised that the change is reverted at some point, cause TiDB is designed to be stateless and frequent restart is allowed to happen.

Reply:
This is almost more a question of what should be a candidate for instance scope. Eason's opinion is that we should always default to global scope for management simplicity, and actually few things should be instance. I think CRDB gets this right in that the node specific options almost always refer to local resources (max memory, tmpdir path, log file path, port, socket). We probably get a few things wrong currently, but are mostly on the right path. 

If we look from these examples: you might need to shuffle around the `tmpdir` on a particular TiDB server instance to keep it up and running (and might not want to restart it). So you would have to log into the TiDB instance to make this change, but it would not affect users running through a LB.

As to the semantic of "reverted", if we were to persist the change, it is not really any less confusing because there is still a precedence problem of config file vs. saved in TiDB (i.e. "I set this option in the config file but it's not working!"). The semantic is also how MySQL works (although there is a newer feature called SET PERSIST), TiDB is special in that it persists all global config.

Reply:
I have added a section on "Documentation Changes", PTAL. I think this is easier to explain

Reply:
Thanks for the supply! Does this mean that, we will have a clear rule that for these settings, an INSTANCE variable can be used to temporarily updating the config (and taking effect immediately), while user should use config file for persistently updating the config (but taking effect after a reboot)? 

I do like CRDB's config mechanism a lot. There aren't a lot of ways like TiDB to configure. They only have two, which is very easy for user to learn:

a. CLI Args - Instance level, restart to take effect, persisted
b. Variable - Cluster level, take effect immediately, persisted

Reply:
> Notice that under CRDB's model it is not possible to adjust some instance level things and take effect without a restart.

Yes exactly. The other limitation is explaining why 2 nodes are behaving differently: in MySQL `SHOW GLOBAL VARIABLES` captures all configuration, but since memory in CRDB is a node-level option, it wouldn't be captured, but it is in this proposal since everything has a sysvar mapping (step 2).

> Does this mean that, we will have a clear rule that for these settings, an INSTANCE variable can be used to temporarily updating the config (and taking effect immediately), while user should use config file for persistently updating the config (but taking effect after a reboot)?

Yes, I propose in documentation we describe the semantic as "Persists to Cluster: Bool" (instead of `INSTANCE` vs. `GLOBAL`). `Persists to Cluster: No` is identical to how MySQL global variables, so it is easy to explain. We do not need to use the term Instance scope vs. global scope for users - since they are both set with `SET GLOBAL`.